### PR TITLE
website: Implement RFC MKTG-033

### PR DESCRIPTION
**Pull request by zchsh**
🎟️  [Asana task](https://app.asana.com/0/1100423001970639/1200017441815806/f)
🔍 [Deploy preview](https://packer-git-fork-zchsh-zsremote-plugin-zip-approach-hashicorp.vercel.app/docs/builders/docker)

This PR implements the [MKTG-033 RFC](https://docs.google.com/document/d/1lfObyzbI_nFPB-FICDPwsoGa4sk323tfJNLnIdLXZUA/). This switches our consumption method for plugin documentation to consume `docs.zip` release assets.

The main shift in this approach is that we no longer use the GitHub API to fetch plugin docs, so we  longer need a `GITHUB_API_TOKEN` to build the site. Further details can be found in [the MKTG-033 RFC](https://docs.google.com/document/d/1lfObyzbI_nFPB-FICDPwsoGa4sk323tfJNLnIdLXZUA/).

This PR is intended as a replacement for #10716.

## Summary of changes

- **Switches to fetch plugin docs from release assets**
    - Removes the need for a GitHub token
    - Plugin docs configuration entries in `website/data/docs-remote-plugins.json` must now include a `version` property
    - At build time, we fetch a `docs.zip` asset from the tagged release that matches the configured `version`
    - Where `version` is not set to `"latest"`, the build automatically falls back to fetching the source `{version}.zip` and parsing docs from that file instead. This PR demonstrates this fallback by fetching `packer-plugin-docker` docs from `v0.0.4` ([preview](https://packer-git-fork-zchsh-zsremote-plugin-zip-approach-hashicorp.vercel.app/docs/builders/docker)) ([v0.0.4 release](https://github.com/hashicorp/packer-plugin-docker/releases/tag/v0.0.4))
- **Adds search indexing for plugin docs**
    - Plugin docs will now be searchable via the bar at the top of each docs page (once we merge into `stable-website` and indexing completes)
    - Note: we have [a broader platform task to revisit how we approach indexing docs content](https://app.asana.com/0/1100423001970639/1200012330686466/f) across all product sites.
    - We've tested the current approach in this PR with [a staging index in Algolia](https://www.algolia.com/apps/YY0FFNI7MF/explorer/browse/staging_PACKER). 
- **Updates README**
    - Adds notes on managing remote plugin docs
    - Updates to reflect the revised `nav-data.json` format from [MKTG-032](https://docs.google.com/document/d/1kYvbyd6njHFSscoE1dtDNHQ3U8IzaMdcjOS0jg87rHg/edit)
- Pulls in a stable version of `docs-page`
- Fixes minor style issue with `<PluginTierLabel />`

### CODE REVIEW

File: website/README.md
Comment (jescalan) [inline website/README.md:68]: Is this still the case?

File: website/README.md
Comment (jescalan) [inline website/README.md:391]: good idea for the docs to allow comments this way 👀

**Review by jescalan — COMMENTED** [review]
Looking great so far!

File: website/README.md
Comment (zchsh) [reply website/README.md]: No, good catch! Have now removed this section, previewing plugin docs "just works" now 👍

File: website/package.json
Comment (zchsh) [inline website/package.json:20]: `adm-zip` is used to unzip fetched `docs.zip` (and plugin source fallbacks).

File: website/package.json
Comment (zchsh) [inline website/package.json:21]: `gray-matter` is used to parse `nav_title` or `sidebar_title` values from `docs.mdx` files, so we can construct `nav-data` for plugin docs with human-readable titles (rather than just using `.mdx` file basenames).

File: website/pages/guides/[[...page]].jsx
Comment (zchsh) [inline website/pages/guides/[[...page]].jsx:18]: Changes to the `docs`, `guides`, and `intro` page routes reflect the slightly revised API in the stable release of `docs-page`.

File: website/scripts/index_search_content.js
Comment (zchsh) [inline website/scripts/index_search_content.js:11]: This indexing approach uses some existing and slightly modified search utilities, which we exposed via https://github.com/hashicorp/react-components/pull/161, to allow us to index content without messing with the filesystem (as we did in our initial approach).

I've testing this indexing with [a temporary Algolia staging index for packer](https://www.algolia.com/apps/YY0FFNI7MF/explorer/browse/staging_PACKER).

> Side note: I would love to get a staging index setup for deploy previews, so we can better test the indexing changes... This is something we have on our `learn` site)... but that's an effort for another day! Going to add this idea to our web platform board.

File: website/README.md
Comment (nywilken) [inline website/README.md:385]:
```suggestion
The `docs-remote-plugins.json` file contains an array of entries. Each entry points to a plugin repository. The `{ title, path, repo, version }` properties are required for each entry, all other properties are optional.
```

File: website/components/remote-plugin-docs/utils/fetch-plugin-docs.js
Comment (nywilken) [inline website/components/remote-plugin-docs/utils/fetch-plugin-docs.js:25]:
```suggestion
    const err = `Failed to fetch. Could not find "docs.zip" at ${docsZipUrl}. To fall back to parsing docs from "source", please provide a specific version tag instead of "${tag}".`
```

File: website/components/remote-plugin-docs/utils/validate-plugin-docs-files.js
Comment (nywilken) [reply website/components/remote-plugin-docs/utils/validate-plugin-docs-files.js:14]: Great idea :+1:

File: website/data/docs-remote-plugins.json
Comment (nywilken) [reply website/data/docs-remote-plugins.json:6]: Nice - I will remove the .docs-artifacts directory in the next PR to packer-plugin-docker. The release v0.0.6 now has a docs.zip

**Review by nywilken — APPROVED** [review]
This looks great. Nicely done. I left two suggestions around version and tag. You use the interchangeably but I believe the docs-remote-navigation.json file is expected version so I left a suggested change there. But this is otherwise good. 

I open a PR on the scaffolding project to reflect this change. Please let me know if it is correct https://github.com/hashicorp/packer-plugin-scaffolding/pull/26

**Comment by zchsh** [comment]
Thanks  @nywilken! I took a look at https://github.com/hashicorp/packer-plugin-scaffolding/pull/26, just had a couple nitpicks in the README. I'll wait for a final review from @jescalan here, and then will merge away 🎉 🚢

File: .github/workflows/linkchecker.yml
Comment (zchsh) [inline .github/workflows/linkchecker.yml:10]: Bumped this since builds are hovering around 3 minutes, so I was getting some flaky CI passes.

File: .github/workflows/check-plugin-docs.yml
Comment (sylviamoss) [inline .github/workflows/check-plugin-docs.yml:14]: As soon as a plugin present in `website/data/docs-remote-plugins.json` releases a new latest version with invalid docs, all PRs with changes in `website/*` will start failing because of that. 

If it's possible, I think it makes sense to have it scheduled to regularly check remote docs and also on:
```
on:
  pull_request:
    paths:
      - "website/data/docs-remote-plugins.json"
```

File: website/README.md
Comment (sylviamoss) [inline website/README.md:396]: I remember a conversation about making partner's plugins official as well but I couldn't find the conversation and the result of it. In case that came to happen in the future, I suppose this will be easy to change, right?

File: .github/workflows/check-plugin-docs.js
Comment (sylviamoss) [inline .github/workflows/check-plugin-docs.js:19]: if one opens a PR with an invalid configuration, for example missing the version, what happens to this script? 
Do you think it makes sense to quickly validate, in this same script, that everything is there (title, path, repo, and version). Don't need to implement it if you think is not a good idea 🤔

**Review by sylviamoss — APPROVED** [review]
This is nice! I added some questions/suggestions other than that LGTM! 👍🏼

File: website/components/remote-plugin-docs/server.js
Comment (jescalan) [inline website/components/remote-plugin-docs/server.js:35]: How does this change after the zip approach, since we no longer have artifact files?

File: website/components/remote-plugin-docs/utils/parse-source-zip.js
Comment (jescalan) [inline website/components/remote-plugin-docs/utils/parse-source-zip.js:49]: I'm slightly confused by `parse-source-zip` and `parse-docs-zip` -- they appear to be very similar in several ways but slightly different. Aren't we only taking in a single zip file?

Edit: Never mind, saw the comment below explaining this. Might be helpful to add some comments to the files themselves up top!

File: website/pages/guides/[[...page]].jsx
Comment (jescalan) [reply website/pages/guides/[[...page]].jsx:18]: Was the `mainBranch` prop removed on purpose? Just want to double check

**Review by jescalan — APPROVED** [review]
Looking good! Approved pending replies to comments 😀

File: .github/workflows/check-plugin-docs.yml
Comment (zchsh) [reply .github/workflows/check-plugin-docs.yml:14]: @sylviamoss I think the scheduled check would be a great idea - definitely open to adding that in a follow up PR, or maybe even here if it's straightforward (I need to brush up on my GH actions and see what's possible, will report back here once I have a rough idea 🏃) 

Re: changing the PR trigger for this check, I've been thinking of this "plugins" check as a kind of "canary" for the Vercel deploy check. If there are any invalid plugin docs, (eg due to a new latest version as in the scenario you mentioned), then all Vercel deploy previews, on all PRs that touch `website/*`, will start failing... but only folks with access (and time) to dive into the Vercel build logs will be able to figure out why. So I think it's preferable to have this check run every time we try to build the site, not just when the plugin config changes. That way, if the Vercel build fails, there's a quick indicator of whether or not it's related to invalid plugin docs.

Let me know if that makes sense - also more than happy to adjust, I feel like I might be missing something, but just wanted to make sure the intention was clear 🙇

File: website/README.md
Comment (zchsh) [reply website/README.md:396]: @sylviamoss great callout, yeah this should be relatively easy to adjust in the future - eg we could make the `official` status directly configurable in `docs-remote-plugins.json`, and fallback to inferring it from the org 👍

File: .github/workflows/check-plugin-docs.js
Comment (zchsh) [reply .github/workflows/check-plugin-docs.js:19]: I think this is a great idea! 💖 Will jump on that now, I think it should be relatively straightforward 👍

File: website/components/remote-plugin-docs/server.js
Comment (zchsh) [reply website/components/remote-plugin-docs/server.js:35]: With the ZIP approach as proposed in the RFC, we have the requirements that 1) the `docs` folder lives at the root of the plugin repo and 2) the `docs.zip` is just a `.zip` of said folder. So we can easily infer the original source file that we link to.

That being said I definitely slipped up here, this comment feels a bit out-of-date. I think it might also make sense to move the URL construction back out here for consistency - I'll adjust this 👍

File: website/components/remote-plugin-docs/utils/parse-source-zip.js
Comment (zchsh) [reply website/components/remote-plugin-docs/utils/parse-source-zip.js:49]: Great call, will add them comments 👍

File: website/pages/guides/[[...page]].jsx
Comment (zchsh) [reply website/pages/guides/[[...page]].jsx:18]: It was on purpose! Previously `docs-page` was using it to construct the GitHub `Edit this page` link, but that was all. Now with the flexibility needed to construct a GitHub edit link not based on branch but instead on version, it felt logical to switch up the interface. (see also [related comment on react-components PR](https://github.com/hashicorp/react-components/pull/154/files#r588828157)

File: website/components/remote-plugin-docs/utils/parse-source-zip.js
Comment (zchsh) [reply website/components/remote-plugin-docs/utils/parse-source-zip.js:56]: ✅

File: website/components/remote-plugin-docs/server.js
Comment (zchsh) [reply website/components/remote-plugin-docs/server.js]: ✅ Adjusted the comments here, to reflect that we grab the `sourceUrl` off the `remoteFile` object.

File: .github/workflows/check-plugin-docs.js
Comment (zchsh) [reply .github/workflows/check-plugin-docs.js:19]: Implemented a basic validation check, with a dedicated error message, before trying to fetch. Here's some sample output from a fail due to a missing `version`:

CleanShot 2021-03-20 at 21 47 33@2x

Let me know if this is approx what you had in mind ☺️

File: .github/workflows/check-plugin-docs.yml
Comment (zchsh) [reply .github/workflows/check-plugin-docs.yml:14]: Oh nice it looks fairly simple to [add multiple event types](https://docs.github.com/en/actions/reference/events-that-trigger-workflows#schedule), and looks like we already have a `scheduled` action running to check links, so this seems very very doable. Will update in a sec to run on both PRs and on a recurring basis 👍

Comment (zchsh) [reply .github/workflows/check-plugin-docs.yml:14]: ✅ All done! I've set the plugin docs check to run daily at `00:45`, to match the broken links checked. Happy to adjust if there's a different timing we'd prefer 👍

File: .github/workflows/check-plugin-docs.js
Comment (sylviamoss) [reply .github/workflows/check-plugin-docs.js:19]: This looks great! 👏🏼

File: .github/workflows/check-plugin-docs.yml
Comment (sylviamoss) [reply .github/workflows/check-plugin-docs.yml:14]: Hey, this looks good! I think what you said makes totally sense 👍🏼

File: website/README.md
Comment (sylviamoss) [reply website/README.md:396]: Nice! Thank you!

**Pull request merged**
